File: LeetCode/minimum_Path_Sum.java
/**
 * Given a m x n grid filled with non-negative numbers, find a path from top
 * left to bottom right, which minimizes the sum of all numbers along its path.
 * 
 * Note: You can only move either down or right at any point in time.
 * 
 * 
 * 
 * Example 1:
 * 
 * 
 * Input: grid = [[1,3,1],[1,5,1],[4,2,1]]
 * Output: 7
 * Explanation: Because the path 1 → 3 → 1 → 1 → 1 minimizes the sum.
 * Example 2:
 * 
 * Input: grid = [[1,2,3],[4,5,6]]
 * Output: 12
 * 
 * 
 * Constraints:
 * 
 * m == grid.length
 * n == grid[i].length
 * 1 <= m, n <= 200
 * 0 <= grid[i][j] <= 100
 */
public class minimum_Path_Sum {
    public int minPathSum(int[][] grid) {
        return spaceOptimization(grid);
    }

    private int solve(int[][] grid, int m, int n, int[][] dp) {
        if (m < 0 || n < 0)
            return Integer.MAX_VALUE;

        if (m == 0 && n == 0) {
            return grid[m][n];
        }

        if (dp[m][n] != -1)
            return dp[m][n];

        int left = solve(grid, m, n - 1, dp);
        int up = solve(grid, m - 1, n, dp);

        int currMin = Math.min(left, up);

        return dp[m][n] = grid[m][n] + currMin;
    }

    private int tabulation(int[][] grid){
        int m = grid.length;
        int n = grid[0].length;

        int[][] dp = new int[m][n];

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if(i==0 && j==0) dp[i][j]=grid[i][j];
                else{
                    int left = Integer.MAX_VALUE;
                    int up = Integer.MAX_VALUE;

                    if(i>0){
                        up = dp[i-1][j];
                    }
                    if(j>0){
                        left = dp[i][j-1];
                    }
                    dp[i][j] = grid[i][j] + Math.min(up, left);
                }
            }
        }
        return dp[m-1][n-1];
    }

    private int spaceOptimization(int[][] grid){
        int m = grid.length;
        int n = grid[0].length;

        int[] so = new int[n];

        for (int i = 0; i < m; i++) {
            int[] curr = new int[n];
            for (int j = 0; j < n; j++) {
                if(i==0 && j==0) curr[j]=grid[i][j];
                else{
                    int left = Integer.MAX_VALUE;
                    int up = Integer.MAX_VALUE;

                    if(i>0){
                        up = so[j];
                    }
                    if(j>0){
                        left = curr[j-1];
                    }
                    curr[j] = grid[i][j] + Math.min(up, left);
                }
            }
            so = curr;
        }
        return so[n-1];
    }
}

File: Leetcode-solution-c-c++/minimumPathSum.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
private:
    int recursion(int x, int y, vector<vector<int>> &grid, int m, int n)
    {
        if (x < 0 || y < 0)
        {
            return int(4e6);
        }
        if (x == 0 && y == 0)
        {
            return grid[0][0];
        }
        int left = recursion(x - 1, y, grid, m, n);
        int up = recursion(x, y-1, grid, m, n);
        return grid[x][y] + min(left, up);
    }

    int memoization(int x, int y, vector<vector<int>> &grid, int m, int n, vector<vector<int>> &dp)
    {
        if (x < 0 || y < 0)
        {
            return int(4e6);
        }
        if (x == 0 && y == 0)
        {
            return grid[0][0];
        }
        if(dp[x][y] != -1) 
        {
            return dp[x][y];
        }
        int left = memoization(x - 1, y, grid, m, n, dp);
        int up = memoization(x, y-1, grid, m, n, dp);
        return dp[x][y] = grid[x][y] + min(left, up);
    }

    int tabulation(vector<vector<int>> &grid, int m, int n)
    {
        vector<vector<int>> dp(m+1, vector<int> (n+1, 0));
        for(int i=0; i<=m; i++) dp[i][0] = 4e6;
        for(int i=0; i<=n; i++) dp[0][i] = 4e6;
        dp[1][1] = grid[0][0];
        
        for(int i=1; i<=m; i++){
            for(int j=1; j<=n; j++){
                int left = dp[i-1][j];
                int up = dp[i][j-1];
                dp[i][j] = grid[i-1][j-1] + min(left, up);
            }
        }

        return dp[m][n];
    }

    int spaceOptimization(vector<vector<int>> &grid, int m, int n)
    {
        vector<int> prev(n+1, 0);
        for(int i=2; i<=n; i++) prev[i] = 4e6;
        
        for(int i=1; i<=m; i++){
            vector<int> curr(n+1, 0);
            curr[0] = 4e6;
            for(int j=1; j<=n; j++){
                int left = prev[j];
                int up = curr[j-1];
                curr[j] = grid[i-1][j-1] + min(left, up);
            }
            prev = curr;
        }

        return prev[n];
    }

public:
    int minPathSum(vector<vector<int>> &grid)
    {
        int m = grid.size();
        int n = grid[0].size();
        return spaceOptimization(grid, m, n);
        // vector<vector<int>> dp(m, vector<int>(n, -1));
        // return memoization(m - 1, n - 1, grid, m, n, dp);
    }
};